ERCS/erc-7739.md

```diff
@@ -0,0 +1,651 @@
+/// This method intentionally returns bytes32 to allow freedom for future extensions.
+function supportsNestedTypedDataSign() public view virtual returns (bytes32 result) {
+    result = bytes4(0xd620c85a);
+}
+```
+
+### Workflow detection
+
+The `isValidSignature` function MUST perform the `TypedDataSign` workflow if the signature contains the correct data to reconstruct the `hash`.
+
+This can be checked via:
+
+```solidity
+// If this is true, it means that the `signature` contains 
+// the correct `APP_DOMAIN_SEPARATOR` and `contents`.
+hash == keccak256(
+    abi.encodePacked(
+        hex"1901",
+        bytes32(APP_DOMAIN_SEPARATOR),
+        bytes32(contents)
+    )
+)
+```
+
+If this expression returns false, then the `PersonalSign` workflow MUST be attempted.
+
+### Conditional skipping of defensive rehashing
+
+Smart accounts MAY skip the defensive rehashing workflows if any of the following is true:
+
+- `isValidSignature` is called off-chain.
+- The `hash` passed into `isValidSignature` has already included the address of the smart account.
+
+## Rationale
+
+### `TypedDataSign` structure
+
+The `typedDataSignTypehash` must be constructed on-the-fly on-chain. This is to enforce that the signed contents will be visible in the signature request, by requiring that `contents` be a user defined type. 
+
+The structure is intentionally made flat with the fields of `eip712Domain` to make implementation feasible. Otherwise, smart accounts must implement on-chain lexographical sorting of strings for the struct type names when constructing `typedDataSignTypehash`.
+
+### `supportsNestedTypedDataSign` for detection
+
+Without this function, this standard will not change the interface of the smart account, as it defines the behavior of `isValidSignature` without adding any new functions. As such, [ERC-165](./eip-165.md) cannot be used.
+
+For future extendability, `supportsNestedTypedDataSign` is defined to return a bytes32 as the first word of its returned data. For bytecode compactness and to leave space for bit packing, only the leftmost 4 bytes are set to the function selector of `supportsNestedTypedDataSign`.
+
+The `supportsNestedTypedDataSign` function may be extended to return multiple values (e.g. `bytes32 result, bytes memory data`), as long as the first word of the returned data is a bytes32 identifier. This will not change the function selector.
+
+## Backwards Compatibility
+
+No backwards compatibility issues.
+
+## Reference Implementation
+
+The following reference implementation is production ready and optimized. It also includes relevant complementary features required for safety, flexibility, developer experience, and user experience.
+
+It is intentionally not minimalistic. This is to avoid repeating the mistake of [ERC-1271](./eip-1271.md), where a reference implementation is wrongly assumed to be safe for production use.
+
+```solidity
+// SPDX-License-Identifier: MIT
```

Comment:
Code within the EIP itself must be CC0-1.0 licensed. If you'd like, you can include MIT code in the assets directory and link to it.